service/history/shard/controller.go

```diff
@@ -299,22 +303,28 @@ func (c *controller) getOrCreateHistoryShardItem(shardID int) (*historyShardsIte
 	return nil, CreateShardOwnershipLostError(c.GetHostInfo().Identity(), info.GetAddress())
 }
 
-func (c *controller) removeHistoryShardItem(shardID int) (*historyShardsItem, error) {
+func (c *controller) removeHistoryShardItem(shardID int, shardItem *historyShardsItem) (*historyShardsItem, error) {
 	nShards := 0
 	c.Lock()
-	shardItem, ok := c.historyShards[shardID]
+	defer c.Unlock()
+
+	currentShardItem, ok := c.historyShards[shardID]
 	if !ok {
-		c.Unlock()
 		return nil, fmt.Errorf("No item found to remove for shard: %v", shardID)
 	}
+	if shardItem != nil && currentShardItem != shardItem {
+		// the shardItem comparison is a defensive check to make sure we are deleting
+		// what we intend to delete.
+		return nil, fmt.Errorf("Current shardItem doesn't match the one we intend to delete for shard: %v", shardID)
+	}
+
 	delete(c.historyShards, shardID)
 	nShards = len(c.historyShards)
-	c.Unlock()
 
 	c.metricsScope.IncCounter(metrics.ShardItemRemovedCounter)
 
-	shardItem.logger.Info("", tag.LifeCycleStopped, tag.ComponentShardItem, tag.Number(int64(nShards)))
-	return shardItem, nil
+	currentShardItem.logger.Info("", tag.LifeCycleStopped, tag.ComponentShardItem, tag.Number(int64(nShards)))
```

Comment:
is that possible currentShardItem be nil? I saw check on L246

Reply:
It won't be nil here. The check on L246 is for the cases where this function returns an error (L313 and L318)